modules/runtime-common/src/main/scala/com/snowplowanalytics/snowplow/runtime/processing/BatchUp.scala

```diff
@@ -93,62 +95,80 @@ object BatchUp {
    * @return
    *   A FS2 Pipe which combines small `A`s to bigger `A`s.
    */
-  def noTimeout[F[_]: Sync, A: Batchable](maxWeight: Long): Pipe[F, A, A] = {
-    def go(stream: Stream[F, A], unflushed: Option[A]): Pull[F, A, Unit] =
+  def noTimeout[F[_]: Sync, A, B: Batchable[A, *]](maxWeight: Long): Pipe[F, A, B] = {
+    def go(stream: Stream[F, A], unflushed: PendingBatch[B]): Pull[F, B, Unit] =
       stream.pull.uncons.flatMap {
         case None =>
           // Upstream finished cleanly. Emit whatever is pending and we're done.
-          Pull.outputOption1[F, A](unflushed) *> Pull.done
+          Pull.outputOption1[F, B](unflushed.value) *> Pull.done
         case Some((chunk, next)) =>
-          val CombineByWeightResult(notAtSize, toEmit) = combineByWeight(maxWeight, Chunk.fromOption(unflushed) ++ chunk)
-          Pull.output[F, A](Chunk.from(toEmit)) *>
-            go(next, notAtSize)
+          val CombineByWeightResult(doNotEmitYet, pendingWeight, toEmit) = combineByWeight(maxWeight, unflushed, chunk)
+          Pull.output[F, B](Chunk.from(toEmit)) *>
+            go(next, PendingBatch(doNotEmitYet, pendingWeight))
       }
 
-    in => go(in, None).stream
+    in => go(in, PendingBatch.empty).stream
+  }
+
+  private case class PendingBatch[B](value: Option[B], weight: Long)
+
+  private object PendingBatch {
+    def empty[B]: PendingBatch[B] = PendingBatch(None, 0L)
   }
 
   /**
    * The result of combining a chunk of `A`s, while not exceeding total weight.
    *
    * @param notAtSize
-   *   Optionally an `A` that does not yet exceed the maximum allowed size. We should not emit this
-   *   `A` but instead wait in case we can combine it with other `A`s later.
+   *   Optionally a batch `B` that does not yet exceed the maximum allowed size. We should not emit
+   *   this `B` but instead wait in case we can combine it with other `A`s later.
    * @param toEmit
-   *   The combined `A`s which meet size requirements. These should be emitted downstream because we
+   *   The combined `B`s which meet size requirements. These should be emitted downstream because we
    *   cannot combine them with anything more.
    */
-  private case class CombineByWeightResult[A](notAtSize: Option[A], toEmit: Vector[A])
+  private case class CombineByWeightResult[B](
+    doNotEmitYet: Option[B],
+    pendingWeight: Long,
+    toEmit: Vector[B]
+  )
```

Comment:
I find it a bit hard to follow to have both a `PendingBatch` and a `CombineByWeightResult` which is supposed to be `The result of combining a chunk` but that can nevertheless be incomplete

Reply:
Good point.  I re-worked the classes and return types a bit, which I think makes it neater now.